esp-hal/src/dma/buffers.rs

```diff
@@ -4,25 +4,334 @@ use core::{
+                Self {
+                    external_memory,
+                    internal_memory: InternalBurstConfig::DEFAULT,
+                }
+            }
+        }
+    } else {
+        /// Burst transfer configuration.
+        #[derive(Clone, Copy, PartialEq, Eq, Debug)]
+        #[cfg_attr(feature = "defmt", derive(defmt::Format))]
+        pub enum BurstConfig {
+            /// Burst mode is disabled.
+            Disabled,
+
+            /// Burst mode is enabled.
+            Enabled,
+        }
+
+        impl BurstConfig {
+            /// The default burst mode configuration.
+            pub const DEFAULT: Self = Self::Disabled;
+        }
+
+        impl Default for BurstConfig {
+            fn default() -> Self {
+                Self::DEFAULT
+            }
+        }
+
+        type InternalBurstConfig = BurstConfig;
+    }
+}
+
+#[cfg(psram_dma)]
+impl ExternalBurstConfig {
+    const fn min_psram_alignment(self, direction: TransferDirection) -> usize {
+        // S2 TRM: Specifically, size and buffer address pointer in receive descriptors
+        // should be 16-byte, 32-byte or 64-byte aligned. For data frame whose
+        // length is not a multiple of 16 bytes, 32 bytes, or 64 bytes, EDMA adds
+        // padding bytes to the end.
+
+        // S3 TRM: Size and Address for IN transfers must be block aligned. For receive
+        // descriptors, if the data length received are not aligned with block size,
+        // GDMA will pad the data received with 0 until they are aligned to
+        // initiate burst transfer. You can read the length field in receive descriptors
+        // to obtain the length of valid data received
+        if matches!(direction, TransferDirection::In) {
+            self as usize
+        } else {
+            // S2 TRM: Size, length and buffer address pointer in transmit descriptors are
+            // not necessarily aligned with block size.
+
+            // S3 TRM: Size, length, and buffer address pointer in transmit descriptors do
+            // not need to be aligned.
+            1
+        }
+    }
+}
+
+impl InternalBurstConfig {
+    pub(super) const fn is_burst_enabled(self) -> bool {
         !matches!(self, Self::Disabled)
     }
+
+    // Size and address alignment as those come in pairs on current hardware.
+    const fn min_dram_alignment(self, direction: TransferDirection) -> usize {
+        if matches!(direction, TransferDirection::In) {
+            // NOTE(danielb): commenting this check is incorrect as per TRM, but works.
+            //                we'll need to restore this once peripherals can read a
+            //                different amount of data than what is configured in the
+            //                buffer.
+            // if cfg!(esp32) {
+            //     // NOTE: The size must be word-aligned.
+            //     // NOTE: The buffer address must be word-aligned
+            //     4
```

Comment:
What a mess. Since this check is used by `set_length` indirectly, restoring 4 would mean we reject all transfer lengths that are not multiples of 4. But the hardware seems to contradict the TRM in that it works just fine...

Reply:
What behavior would you expect to see if the TRM wasn't being contradicted?